### DISCUSSION TITLE
Dynamic Email Filtering Using AI (for new Rules System)

### DISCUSSION THREAD

**Discussion by vinnydis** [anonaddy/anonaddy#640] [Feature Requests]
Happy to see the implementation of the Rules System. I'd like to suggest an enhancement to further improve our email management:

* **Issue:**
  - Static rules struggle to adapt to the variable formats of emails and newsletters.
  - Manually updating rules for emails is too time-consuming (takes more time to set up and maintain then to just delete those emails from my inbox).

* **Proposed Solution:**
  - Integrate a Large Language Model (LLM) like OpenAI's GPT to understand and act on user-defined preferences expressed in natural language.
  
* **Example Use Case for Kiwi.com, a travel website I book with frequently and that also sends useless marketing:**
  - **Desired:** Emails with booking confirmations, schedule changes, or delays from Kiwi.com should be allowed.
  - **Block:** Emails promoting deals, hotel offers, or tours should be filtered out.

* **Implementation Idea:**
  - Users specify their email preferences in simple sentences.
  - Addy leverages GPT to evaluate each email's content, deciding to either "PASS" or "BLOCK" based on these preferences.
  - The AI errs on the side of "PASS" when unsure, minimizing the chance of important emails being mistakenly blocked.

An example of how I would state this as a user for Kiwi.com:

> "Please allow emails related to travel updates such as booking confirmations, cancellations, and notifications about delays or schedule changes. Block all promotional emails, especially those offering hotels or activities."

This way, we could manage emails better without fiddling with rules every other day. With the current pricing of these LLMs, this would probably come out to around 50 cents or so per month, and I think many of us would be more than happy to pay for that.

**Comment by beedell-roke** [comment]
@vinnydis, running a vaguely competent LLM (which might have to be manually trained due to this niche application, and prevalence of copyright-infringing datasets) might be too computationally expensive for the official servers. I can imagine it requiring an additional subscription. Though, for those self-hosting, it could work quite well.

**Reply by vinnydis** [reply]
@RokeJulianLockhart Thanks for your thoughts on this! Just wanted to mention that I’ve actually tested this using the publicly available OpenAI GPT-4o API with a simple prompt and the body of a few emails, and it works quite reliably for filtering emails as described. Given that the GPT-4o API is fairly cheap, I believe Addy could likely integrate this without too much concern on cost.

That said, it's worth noting that the API is usage-based, so for users with heavier email traffic, costs can accumulate. In my case, with pretty heavy email usage, it would come out to around 20 cents per month—not a big deal, but something to consider for Addy.

A potential solution might be to allow users to plug in their own OpenAI API credentials, so they pay for their own usage. Alternatively, Addy could charge a small additional fee for those who prefer a fully integrated solution. What do you think?

**Reply by beedell-roke** [reply]
@vinnydis, an integration system, designed to be modular enough to support any arbitrary API (so-to-speak) — with something like GPT-4o implemented by default — would be ideal, because if the user uses that premade integration, all they need is an API key. I think a key and 20 cents/m is easily accessible enough for most.
